### PR TITLE
feat(ci): migrate to github actions

### PR DESCRIPTION
**Pull request by Crow-EH**
Only a simple draft of migration to GitHub Actions. (EDIT: not a draft anymore)
I won't lie, it's mainly for keeping myself busy between christmas and new year's eve. 🎅 
No problem if it's trashed in the end :p.
Feedbacks are welcome, and happy new year !

PS: The matrix was awfully undocumented, but I think I managed to find an ok solution.
PS2: In the future, I'd like to do a separate, conditionnal step only for pushing the images when on `master` and simplify the code. The release could also be an entirely different Action that could simply pull all `latest` from the matrix and promote them to tagged releases instead of rebuilding everything (it seems there's a "deploy" trigger somewhere).
PS3: It seems like jobs are a bit faster on GitHub

Pre-merge TODO for @mremi and/or @axelpavageau:
- [x] Choose the recurence of the scheduled job (midnight UTC at the moment)
- [x] [Setup Docker secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets#creating-encrypted-secrets) DOCKER_USERNAME and DOCKER_PASSWORD (Secrets are not passed to workflows that are triggered by a pull request from a fork, are encrypted and are automatically obfuscated in the logs)
- [x] Remove the continuous-integration/travis-ci check requirement and require the GitHub Action checks under `Settings > Branches > Branch protection rule > Require status checks to pass before merging`. Sadly, it seems that we can't require a check on the whole build and have to require every single jobs.

### CODE REVIEW

**Comment by Mallear** [comment]
Nice move !

I don't know who knows the registry credentials, maybe @mremi ?

For the script name, we can go for the generic 'main.py'  or for something like 'image_builder.py'  which will make `python image_builder.py build` which sound clear for me about what the script is and what it does.

**Comment by Crow-EH** [comment]
# Some testing
I tested some points on a forked repository before switching to a non-draft review.
With those out of the way, the remaining CI is the existing one and should work as usual.

## Secrets
I set up two secret variables, `DOCKER_USERNAME` and `DOCKER_PASSWORD`, to check if `secrets.MY_SECRET` was the correct way to get them.

[It turns out that GitHub will automatically obfuscate any secret printed to the logs](https://github.com/Crow-EH/docker-buildbox/runs/387120664#step:5:13), even when reassigned to other variables.

To test my secrets, I had to cheat a bit:
```
if env_conf["docker_reg_username"] == "mySecretUsername" and env_conf["docker_reg_password"] == "mySecretPassword":
        print("The secrets were correctly imported")
```
=> [Secrets are working correctly](https://github.com/Crow-EH/docker-buildbox/runs/387120664#step:5:19)

So @mremi or @axelpavageau will have to create the secrets DOCKER_USERNAME and DOCKER_PASSWORD before merging anything

## Cron Trigger
[The cron is working as intended, and its event name is also the intended `schedule` keyword](https://github.com/Crow-EH/docker-buildbox/runs/387139869#step:5:15)

## Release Trigger
TODO

**Comment by Crow-EH** [comment]
I opened the PR for review since all my tests were successful.

Please do not hesitate to assess my work, ask any question, discuss / vote for the action and job names, print, etc.

**Comment by Mallear** [comment]
Hello @Crow-EH , do you know if it's possible to make conditionnal build depending on which files are modified by the last commit ? As in gitlab CI. That could speed up MR build.

**Comment by saugereau** [comment]
> Hello @Crow-EH , do you know if it's possible to make conditionnal build depending on which files are modified by the last commit ? As in gitlab CI. That could speed up MR build.

Maybe this : https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths ?

**Comment by Crow-EH** [comment]
That's a cool idea ! I'll see what I can do.

**Comment by Crow-EH** [comment]
If we want to do it with what @saugereau shared, we *have* to stop using the matrix system, but also have multiple workflows it seems.

Il try to find if there's a way to do it at jobs level, but anyway it will require a deeper refactoring of our build system.

**Comment by Mallear** [comment]
> If we want to do it with what @saugereau shared, we _have_ to stop using the matrix system, but also have multiple workflows it seems.
> 
> Il try to find if there's a way to do it at jobs level, but anyway it will require a deeper refactoring of our build system.

You mean, the declaration of jobs or also the main script ?
Templating a build job launching `python travis.py build -i {IMAGE} -v {VERSION}` and overriding IMAGE and VERSION vars is something possible with GithubActions ?

**Comment by saugereau** [comment]
Maybe can we add this in a dedicated issue/enhancement instead of use this thread ? This allow to close this PR and use another thread to discuss about it ?

**Comment by Crow-EH** [comment]
Yeah we totally can implement it as is and do this in a separate PR.
As discussed this evening, we could also drop the matrix altogether, then do separated job (or workflows, I still have to check) that directly pass the variables to python.
The matr

**Comment by Crow-EH** [comment]
The branche has been rebased and the configuration is up to date with master.

The TODO list in the description requires rights that I don't have, so it's still impossible to merge this PR.

I also have started to work on the conditional build on, see the PR #271.

**Comment by saugereau** [comment]
``Choose the recurence of the scheduled job (midnight UTC at the moment)``
Currently travis-ci is setup with daily cron at 2 PM, so I think it's good if we setyp github cron daily at midnight (you cannot set them to run at a specific time in travis-ci).

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    types: [published]
+  schedule:
+    - cron: "0 0 * * *"
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
```
[the remark this answers is not in the document]
Comment (Crow-EH) [reply .github/workflows/build.yml]: That's the our current behavior on Travis, but I'll happily remove it if everyone agree.

[the remark this answers is not in the document]
Comment (clemstoquart) [reply .github/workflows/build.yml]: @saugereau Not sure while every jobs are independent.

[the remark this answers is not in the document]
Comment (clemstoquart) [reply .github/workflows/build.yml]: But we can fail-fast, it's ok with me.

[the remark this answers is not in the document]
Comment (Crow-EH) [reply .github/workflows/build.yml]: Removed (fail-fast: true is the default behaviour)

**Pull request merged**